lib/msal-browser/src/cache/BrowserStorage.ts

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ICacheStorage, Constants, PersistentCacheKeys, TemporaryCacheKeys } from "msal-common";
+import { CacheOptions } from "../app/Configuration";
+import { BrowserAuthError } from "../error/BrowserAuthError";
+import { BrowserConfigurationAuthError } from "../error/BrowserConfigurationAuthError";
+import { BrowserConstants } from "../utils/BrowserConstants";
+import { ErrorCacheKeys } from "msal-common/dist/utils/Constants";
+
+export class BrowserStorage implements ICacheStorage {
+
+    private cacheConfig: CacheOptions;
+    private windowStorage: Storage;
+
+    private clientId: string;
+    private rollbackEnabled: boolean;
+
+    constructor(clientId: string, cacheConfig: CacheOptions) {
+        if (!window) {
+            throw BrowserAuthError.createNoWindowObjectError();
+        }
+
+        const storageSupported = typeof window[cacheConfig.cacheLocation] !== "undefined" && window[cacheConfig.cacheLocation] !== null;
+        if (!storageSupported) {
+            throw BrowserConfigurationAuthError.createStorageNotSupportedError();
+        }
+
+        this.cacheConfig = cacheConfig;
+        this.windowStorage = window[this.cacheConfig.cacheLocation];
+        this.clientId = clientId;
+
+        // This is hardcoded to true for now, we will roll this out as a configurable option in the future.
+        this.rollbackEnabled = true;
+
+        this.migrateCacheEntries();
+    }
+
+    /**
+     * Support roll back to old cache schema until the next major release: true by default now
+     * @param storeAuthStateInCookie
+     */
+    private migrateCacheEntries() {
+
+        const idTokenKey = `${Constants.CACHE_PREFIX}.${PersistentCacheKeys.ID_TOKEN}`;
+        const clientInfoKey = `${Constants.CACHE_PREFIX}.${PersistentCacheKeys.CLIENT_INFO}`;
+        const errorKey = `${Constants.CACHE_PREFIX}.${ErrorCacheKeys.ERROR}`;
+        const errorDescKey = `${Constants.CACHE_PREFIX}.${ErrorCacheKeys.ERROR_DESC}`;
+
+        const idTokenValue = this.getItem(idTokenKey);
+        const clientInfoValue = this.getItem(clientInfoKey);
+        const errorValue = this.getItem(errorKey);
+        const errorDescValue = this.getItem(errorDescKey);
+
+        const values = [idTokenValue, clientInfoValue, errorValue, errorDescValue];
+        const keysToMigrate = [PersistentCacheKeys.ID_TOKEN, PersistentCacheKeys.CLIENT_INFO, ErrorCacheKeys.ERROR, ErrorCacheKeys.ERROR_DESC];
+
+        keysToMigrate.forEach((cacheKey, index) => this.createCacheEntry(cacheKey, values[index]));
+    }
+
+    /**
+     * Utility function to help with rollback keys
+     * @param newKey
+     * @param value
+     * @param storeAuthStateInCookie
+     */
+    private createCacheEntry(newKey: string, value: string) {
+        if (newKey && value) {
```

Comment:
I'm not sure you need or want this check. Would `key` ever be falsey? And this would prevent setting `null` or empty string.

Reply:
this is a helper specifically for migratingCacheEntries. I will change the name and comment to reflect this.

Reply:
Done, please re-review and resolve.